README.md
# curibio.sdk
Software development kit for utilizing Curi Bio data files and online databases


CHANGELOG.rst
Changelog for Curi Bio Software Development Kit
===============================================

0.10.1 (2021-01-19)
-------------------
- Bumped Docker Container to 3.9.1-slim-buster
- Added message in Jupyter Notebook if not running the latest version

0.10.0 (2021-01-15)
-------------------
- Added twitch frequencies chart excel sheet.
- Added force frequency relationship chart excel sheet.


0.9.0 (2021-01-06)
------------------
- Added Area Under the Curve metric to per twitch metrics sheet and aggregate metrics sheet
- Fixed issue with interpolation values outside of the given boundaries for optical data.


0.8.2 (2020-12-29)
------------------

- Fixed issue with getting the incorrect well index from the well name for optical data.


0.8.1 (2020-12-20)
------------------

- Added Python 3.9 support.
- Added steps to documentation explaining how to analyze multiple zip files.
- Changed formatting of .xlsx output file names to match input the formatting
  of the input file names. A discrepancy still exists between the input and
  output file names, however.
- Added excel sheet for per twitch metrics.


0.8.0 (2020-11-11)
------------------

- Added excel sheet for full length charts.
- Fixed issue with pure noise files causing errors.


0.7.3 (2020-11-05)
------------------

- Fixed issue with twitches point up field for optical data.
- Fixed case sensitivity issue ('y' and 'Y' both work now).
- Fixed issue causing change of chart bounds to be tedious.
- Fixed Y axis label for optical data (now 'Post Displacement (microns)').
- Fixed many of the issues causing two consecutive relaxations to be
  detected incorrectly.
- Fixed interpolation bugs.
- Fixed documentation issues.
- Changed Sampling / Frame Rate from period in seconds to a rate in Hz.


0.7.1 (2020-10-20)
------------------

- Fixed issue with markers in optical data charts.


0.7.0 (2020-10-15)
------------------

- Added ability to analyze optical data entered in an excel template.
- Added firmware version to excel metadata sheet.


0.6.0 (2020-10-07)
------------------

- Added numbered steps to getting started documentation.
- Added ``contiuous-waveform-plots`` sheet to excel file generation.
  Currently, the only format for chart creation is a <= 10 second "snapshot" of
  the middle data points. It shows waveforms as well as Contraction and
  Relaxation markers on twitches.
- Added access to reference sensor data.
- Added performance improvements for accessing raw data.
- Added ability to upload zip files to Jupyter and updated ``Getting Started``
  documentation to show how to do so.
- Changed all interpolation to 100 Hz.
- Changed default filter for 1600 µs sampling period from Bessel Lowpass 30Hz
  to Butterworth Lowpass 30Hz.
- Fixed peak detection algorithm so it is less likely to report two
  contractions/relaxations of a twitch in a row.


0.5.0 (2020-09-21)
------------------

- Added logging to ``write_xlsx``.
- Added backwards compatibility with H5 file versions >= ``0.1.1``.


0.4.1 (2020-09-16)
------------------

- Added Jupyter getting started documentation.


0.4.0 (2020-09-16)
------------------

- Added support for MyBinder.
- Added Peak Detection Error handling.
- Added function to create stacked plot.


0.3.0 (2020-09-09)
------------------

- Added generation of Excel file with continuous waveform and aggregate metrics.
- Added SDK version number to metadata sheet in Excel file.


tests/utils.py
# -*- coding: utf-8 -*-
def get_cell_value(sheet, zero_based_row, zero_based_col):
    return sheet.cell(row=zero_based_row + 1, column=zero_based_col + 1).value


tests/fixtures.py
# -*- coding: utf-8 -*-
import os
import tempfile

from curibio.sdk import ExcelWellFile
from curibio.sdk import PlateRecording
from curibio.sdk import WellFile
from mantarray_waveform_analysis import BESSEL_LOWPASS_10_UUID
from mantarray_waveform_analysis import CENTIMILLISECONDS_PER_SECOND
from mantarray_waveform_analysis import PipelineTemplate
import pytest
from stdlib_utils import get_current_file_abs_directory

PATH_OF_CURRENT_FILE = get_current_file_abs_directory()


@pytest.fixture(scope="function", name="generic_well_file_0_3_1")
def fixture_generic_well_file_0_3_1():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "h5",
            "v0.3.1",
            "MA20123456__2020_08_17_145752__B3.h5",
        )
    )
    yield wf


@pytest.fixture(scope="session", name="generic_well_file_0_3_2")
def fixture_generic_well_file_0_3_2():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "h5",
            "v0.3.2",
            "MA20223322__2020_09_02_173919",
            "MA20223322__2020_09_02_173919__A2.h5",
        )
    )
    yield wf


@pytest.fixture(
    scope="session", name="plate_recording_in_tmp_dir_for_generic_well_file_0_3_2"
)
def fixture_plate_recording_in_tmp_dir_for_generic_well_file_0_3_2(
    generic_well_file_0_3_2,
):
    period = (
        generic_well_file_0_3_2.get_tissue_sampling_period_microseconds()
        / 1000000
        * CENTIMILLISECONDS_PER_SECOND
    )
    pt = PipelineTemplate(
        noise_filter_uuid=BESSEL_LOWPASS_10_UUID,
        tissue_sampling_period=period,
    )
    pr = PlateRecording([generic_well_file_0_3_2], pipeline_template=pt)
    with tempfile.TemporaryDirectory() as tmp_dir:
        yield pr, tmp_dir


@pytest.fixture(scope="session", name="plate_recording_in_tmp_dir_for_24_wells_0_3_2")
def fixture_plate_recording_in_tmp_dir_for_24_wells_0_3_2():
    pr = PlateRecording.from_directory(
        os.path.join(
            PATH_OF_CURRENT_FILE, "h5", "v0.3.2", "MA20223322__2020_09_02_173919"
        )
    )
    with tempfile.TemporaryDirectory() as tmp_dir:
        yield pr, tmp_dir


@pytest.fixture(scope="session", name="real_3min_well_file_0_3_1")
def fixture_real_3min_well_file_0_3_1():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "h5",
            "v0.3.1",
            "MA201110001__2020_09_03_213024__A3.h5",
        )
    )
    yield wf


@pytest.fixture(
    scope="session", name="plate_recording_in_tmp_dir_for_real_3min_well_file_0_3_1"
)
def fixture_plate_recording_in_tmp_dir_for_real_3min_well_file_0_3_1(
    real_3min_well_file_0_3_1,
):
    period = (
        real_3min_well_file_0_3_1.get_tissue_sampling_period_microseconds()
        / 1000000
        * CENTIMILLISECONDS_PER_SECOND
    )
    pt = PipelineTemplate(
        noise_filter_uuid=BESSEL_LOWPASS_10_UUID,
        tissue_sampling_period=period,
    )
    pr = PlateRecording([real_3min_well_file_0_3_1], pipeline_template=pt)
    with tempfile.TemporaryDirectory() as tmp_dir:
        yield pr, tmp_dir


@pytest.fixture(
    scope="function", name="plate_recording_in_tmp_dir_for_generic_well_file_0_3_1"
)
def fixture_plate_recording_in_tmp_dir_for_generic_well_file_0_3_1(
    generic_well_file_0_3_1,
):
    period = (
        generic_well_file_0_3_1.get_tissue_sampling_period_microseconds()
        / 1000000
        * CENTIMILLISECONDS_PER_SECOND
    )
    pt = PipelineTemplate(
        noise_filter_uuid=BESSEL_LOWPASS_10_UUID,
        tissue_sampling_period=period,
    )
    pr = PlateRecording([generic_well_file_0_3_1], pipeline_template=pt)
    with tempfile.TemporaryDirectory() as tmp_dir:
        yield pr, tmp_dir


@pytest.fixture(scope="function", name="generic_well_file_0_3_1__2")
def fixture_generic_well_file_0_3_1__2():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "h5",
            "v0.3.1",
            "MA20123456__2020_08_17_145752__A2.h5",
        )
    )
    yield wf


@pytest.fixture(
    scope="function", name="plate_recording_in_tmp_dir_for_multiple_well_files_0_3_1"
)
def fixture_plate_recording_in_tmp_dir_for_multiple_well_files_0_3_1(
    generic_well_file_0_3_1, generic_well_file_0_3_1__2
):
    period = (
        generic_well_file_0_3_1.get_tissue_sampling_period_microseconds()
        / 1000000
        * CENTIMILLISECONDS_PER_SECOND
    )
    pt = PipelineTemplate(
        noise_filter_uuid=BESSEL_LOWPASS_10_UUID,
        tissue_sampling_period=period,
    )
    pr = PlateRecording(
        [generic_well_file_0_3_1, generic_well_file_0_3_1__2], pipeline_template=pt
    )
    with tempfile.TemporaryDirectory() as tmp_dir:
        yield pr, tmp_dir


@pytest.fixture(scope="function", name="generic_excel_well_file_0_1_0")
def fixture_generic_excel_well_file_0_1_0():
    file_name = "optical_data_filled_template.xlsx"
    ewf = ExcelWellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "excel_optical_data",
            file_name,
        )
    )
    yield ewf


tests/test_files.py
# -*- coding: utf-8 -*-
import inspect
import os

from curibio.sdk import files
from curibio.sdk import PlateRecording
from curibio.sdk import WellFile
import numpy as np

PATH_OF_CURRENT_FILE = os.path.dirname((inspect.stack()[0][1]))


def test_WellFile__opens_and_get_well_name():
    wf = WellFile(
        os.path.join(PATH_OF_CURRENT_FILE, "h5", "my_barcode__2020_03_17_163600__D6.h5")
    )
    assert wf.get_well_name() == "D6"


def test_WellFile__opens_and_get_well_index():
    wf = WellFile(
        os.path.join(PATH_OF_CURRENT_FILE, "h5", "my_barcode__2020_03_17_163600__D6.h5")
    )
    assert wf.get_well_index() == 23


def test_WellFile__opens_and_get_plate_barcode():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "M120171010__2020_07_22_201922",
            "M120171010__2020_07_22_201922__A1.h5",
        )
    )
    assert wf.get_plate_barcode() == "M120171010"


def test_WellFile__opens_and_get_user_account():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "M120171010__2020_07_22_201922",
            "M120171010__2020_07_22_201922__A1.h5",
        )
    )
    assert wf.get_user_account() == "bab42d5a-25ab-4b88-90ca-55914b55cf58"


def test_WellFile__opens_and_get_customer_account():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "M120171010__2020_07_22_201922",
            "M120171010__2020_07_22_201922__A1.h5",
        )
    )
    assert wf.get_customer_account() == "73f52be0-368c-42d8-a1fd-660d49ba5604"


def test_WellFile__opens_and_get_mantarray_serial_number():
    wf = WellFile(
        os.path.join(
            PATH_OF_CURRENT_FILE,
            "M120171010__2020_07_22_201922",
            "M120171010__2020_07_22_201922__A1.h5",
        )
    )
    assert wf.get_mantarray_serial_number() == "M02001900"


# def test_WellFile__opens_and_get_UTC_begin_recording():
#     wf = WellFile(
#         os.path.join(PATH_OF_CURRENT_FILE, "M120171010__2020_07_22_201922", "M120171010__2020_07_22_201922__A1.h5")
#     )

#     assert wf.get_UTC_begin_recording() == ""


def test_WellFile__opens_and_get_numpy_array():
    wf = WellFile(
        os.path.join(PATH_OF_CURRENT_FILE, "h5", "my_barcode__2020_03_17_163600__D6.h5")
    )
    assert np.size(wf.get_numpy_array()) == 25986


def test_WellFile__opens_and_get_voltage_array():
    wf = WellFile(
        os.path.join(PATH_OF_CURRENT_FILE, "h5", "my_barcode__2020_03_17_163600__D6.h5")
    )
    assert np.size(wf.get_voltage_array()) == 25986


def test_PlateRecording__opens_and_get_wellfile_names():
    wf1 = os.path.join(
        PATH_OF_CURRENT_FILE, "h5_New", "my_barcode__2020_05_24_203716__B1.h5"
    )
    wf2 = os.path.join(
        PATH_OF_CURRENT_FILE, "h5_New", "my_barcode__2020_05_24_203716__B2.h5"
    )
    wf3 = os.path.join(
        PATH_OF_CURRENT_FILE, "h5_New", "my_barcode__2020_05_24_203716__B3.h5"
    )
    wf4 = os.path.join(
        PATH_OF_CURRENT_FILE, "h5_New", "my_barcode__2020_05_24_203716__B4.h5"
    )
    wf5 = os.path.join(
        PATH_OF_CURRENT_FILE, "h5_New", "my_barcode__2020_05_24_203716__B5.h5"
    )
    wf6 = os.path.join(
        PATH_OF_CURRENT_FILE, "h5_New", "my_barcode__2020_05_24_203716__B6.h5"
    )

    file_list = PlateRecording([wf1, wf2, wf3, wf4, wf5, wf6])

    # test csv writer
    file_list.get_combined_csv()

    assert np.size(file_list.get_wellfile_names()) == 6


def test_get_unique_files():
    unique_files = files.get_unique_files_from_directory(PATH_OF_CURRENT_FILE)

    assert len(unique_files) == 76


def test_get_files_by_well_name():
    unique_files = files.get_unique_files_from_directory(
        os.path.join(PATH_OF_CURRENT_FILE, "h5")
    )

    dictionary = files.get_specified_files("Well Name", "D6", unique_files)

    assert len(dictionary["Well Name"]["D6"]) == 1


def test_get_files_by_plate_barcode():
    unique_files = files.get_unique_files_from_directory(
        os.path.join(PATH_OF_CURRENT_FILE, "M120171010__2020_07_22_201922")
    )

    dictionary = files.get_specified_files("Plate Barcode", "M120171010", unique_files)

    assert len(dictionary["Plate Barcode"]["M120171010"]) == 24


def test_get_files_by_user():
    unique_files = files.get_unique_files_from_directory(
        os.path.join(PATH_OF_CURRENT_FILE, "M120171010__2020_07_22_201922")
    )

    dictionary = files.get_specified_files(
        "User ID", "bab42d5a-25ab-4b88-90ca-55914b55cf58", unique_files
    )

    assert len(dictionary["User ID"]["bab42d5a-25ab-4b88-90ca-55914b55cf58"]) == 24


def test_get_files_by_account():
    unique_files = files.get_unique_files_from_directory(
        os.path.join(PATH_OF_CURRENT_FILE, "M120171010__2020_07_22_201922")
    )

    dictionary = files.get_specified_files(
        "Account ID", "73f52be0-368c-42d8-a1fd-660d49ba5604", unique_files
    )

    assert len(dictionary["Account ID"]["73f52be0-368c-42d8-a1fd-660d49ba5604"]) == 24


def test_get_files_by_serial_number():
    unique_files = files.get_unique_files_from_directory(
        os.path.join(PATH_OF_CURRENT_FILE, "M120171010__2020_07_22_201922")
    )

    dictionary = files.get_specified_files(
        "Mantarray Serial Number", "M02001900", unique_files
    )

    assert len(dictionary["Mantarray Serial Number"]["M02001900"]) == 24


prerelease.py
# -*- coding: utf-8 -*-
"""Do the checks and tasks that have to happen before doing a release."""
from __future__ import unicode_literals

import datetime
import logging
import sys

from zest.releaser import baserelease
from zest.releaser import utils

logger = logging.getLogger(__name__)

HISTORY_HEADER = "%(new_version)s (%(today)s)"
PRERELEASE_COMMIT_MSG = "Preparing release %(new_version)s"

# Documentation for self.data.  You get runtime warnings when something is in
# self.data that is not in this list.  Embarrassment-driven documentation!
DATA = baserelease.DATA.copy()
DATA.update(
    {"today": "Date string used in history header",}
)


class Prereleaser(baserelease.Basereleaser):
    """Prepare release, ready for making a tag and an sdist.

    self.data holds data that can optionally be changed by plugins.
    """

    def __init__(self, vcs=None):
        baserelease.Basereleaser.__init__(self, vcs=vcs)
        # Prepare some defaults for potential overriding.
        date_format = self.pypiconfig.date_format()
        self.data.update(
            dict(
                commit_msg=PRERELEASE_COMMIT_MSG,
                history_header=HISTORY_HEADER,
                today=datetime.datetime.today().strftime(date_format),
                update_history=True,
            )
        )

    def prepare(self):
        """Prepare self.data by asking about new version etc."""
        # Tanner (7/29/20): We do not want to do this in GitHub
        # if not utils.sanity_check(self.vcs):
        #     logger.critical("Sanity check failed.")
        #     sys.exit(1)
        if not utils.check_recommended_files(self.data, self.vcs):
            logger.debug("Recommended files check failed.")
            sys.exit(1)
        # Grab current version.
        self._grab_version(initial=True)
        # Grab current history.
        # It seems useful to do this even when we will not update the history.
        self._grab_history()
        if self.data["update_history"]:
            # Print changelog for this release.
            print(
                "Changelog entries for version {0}:\n".format(self.data["new_version"])
            )
            print(self.data.get("history_last_release"))
        # Grab and set new version.
        self._grab_version()
        if self.data["update_history"]:
            # Look for unwanted 'Nothing changed yet' in latest header.
            self._check_nothing_changed()
            # Look for required text under the latest header.
            self._check_required()

    def execute(self):
        """Make the changes and offer a commit."""
        if self.data["update_history"]:
            self._change_header()
        self._write_version()
        if self.data["update_history"]:
            self._write_history()
        self._diff_and_commit()

    def _grab_version(self, initial=False):
        """Grab the version.

        When initial is False, ask the user for a non-development
        version.  When initial is True, grab the current suggestion.
        """
        original_version = self.vcs.version
        logger.debug("Extracted version: %s", original_version)
        if not original_version:
            logger.critical("No version found.")
            sys.exit(1)
        suggestion = utils.cleanup_version(original_version)
        new_version = None
        if not initial:
            new_version = utils.ask_version("Enter version", default=suggestion)
        if not new_version:
            new_version = suggestion
        self.data["original_version"] = original_version
        self.data["new_version"] = new_version


def datacheck(data):
    """Entrypoint: ensure that the data dict is fully documented"""
    utils.is_data_documented(data, documentation=DATA)


def main():
    utils.parse_options()
    utils.configure_logging()
    prereleaser = Prereleaser()
    prereleaser.run()


src/curibio/sdk/__init__.py
# -*- coding: utf-8 -*-
"""Docstring."""

from mantarray_file_manager import WellFile

from .constants import AGGREGATE_METRICS_SHEET_NAME
from .constants import ALL_FORMATS
from .constants import CALCULATED_METRIC_DISPLAY_NAMES
from .constants import CHART_BASE_WIDTH
from .constants import CHART_FIXED_WIDTH
from .constants import CHART_FIXED_WIDTH_CELLS
from .constants import CHART_HEIGHT
from .constants import CHART_HEIGHT_CELLS
from .constants import CHART_WINDOW_NUM_DATA_POINTS
from .constants import CHART_WINDOW_NUM_SECONDS
from .constants import CONTINUOUS_WAVEFORM_SHEET_NAME
from .constants import DEFAULT_CELL_WIDTH
from .constants import EXCEL_OPTICAL_METADATA_CELLS
from .constants import FORCE_FREQUENCY_RELATIONSHIP_SHEET
from .constants import FULL_CHART_SHEET_NAME
from .constants import INTERPOLATED_DATA_PERIOD_CMS
from .constants import INTERPOLATED_DATA_PERIOD_SECONDS
from .constants import INTERPOLATION_VALUE_UUID
from .constants import METADATA_EXCEL_SHEET_NAME
from .constants import METADATA_INSTRUMENT_ROW_START
from .constants import METADATA_OUTPUT_FILE_ROW_START
from .constants import METADATA_RECORDING_ROW_START
from .constants import METADATA_UUID_DESCRIPTIONS
from .constants import MICROSECONDS_PER_CENTIMILLISECOND
from .constants import NUMBER_OF_PER_TWITCH_METRICS
from .constants import PACKAGE_VERSION as __version__
from .constants import PEAK_VALLEY_COLUMN_START
from .constants import PER_TWITCH_METRICS_SHEET_NAME
from .constants import SECONDS_PER_CELL
from .constants import SNAPSHOT_CHART_SHEET_NAME
from .constants import TSP_TO_DEFAULT_FILTER_UUID
from .constants import TWITCH_FREQUENCIES_CHART_SHEET_NAME
from .constants import TWITCHES_POINT_UP_UUID
from .excel_well_file import ExcelWellFile
from .exceptions import MetadataNotFoundError
from .plate_recording import PlateRecording


__all__ = [
    "WellFile",
    "ExcelWellFile",
    "PlateRecording",
    "METADATA_EXCEL_SHEET_NAME",
    "METADATA_RECORDING_ROW_START",
    "METADATA_INSTRUMENT_ROW_START",
    "METADATA_OUTPUT_FILE_ROW_START",
    "CONTINUOUS_WAVEFORM_SHEET_NAME",
    "INTERPOLATED_DATA_PERIOD_CMS",
    "TSP_TO_DEFAULT_FILTER_UUID",
    "MICROSECONDS_PER_CENTIMILLISECOND",
    "CALCULATED_METRIC_DISPLAY_NAMES",
    "AGGREGATE_METRICS_SHEET_NAME",
    "ALL_FORMATS",
    "__version__",
    "SNAPSHOT_CHART_SHEET_NAME",
    "CHART_HEIGHT",
    "CHART_BASE_WIDTH",
    "CHART_HEIGHT_CELLS",
    "PEAK_VALLEY_COLUMN_START",
    "DEFAULT_CELL_WIDTH",
    "CHART_FIXED_WIDTH",
    "CHART_FIXED_WIDTH_CELLS",
    "INTERPOLATED_DATA_PERIOD_SECONDS",
    "CHART_WINDOW_NUM_SECONDS",
    "CHART_WINDOW_NUM_DATA_POINTS",
    "EXCEL_OPTICAL_METADATA_CELLS",
    "TWITCHES_POINT_UP_UUID",
    "METADATA_UUID_DESCRIPTIONS",
    "MetadataNotFoundError",
    "INTERPOLATION_VALUE_UUID",
    "FULL_CHART_SHEET_NAME",
    "SECONDS_PER_CELL",
    "PER_TWITCH_METRICS_SHEET_NAME",
    "NUMBER_OF_PER_TWITCH_METRICS",
    "TWITCH_FREQUENCIES_CHART_SHEET_NAME",
    "FORCE_FREQUENCY_RELATIONSHIP_SHEET",
]


src/curibio/sdk/constants.py
# -*- coding: utf-8 -*-
"""Docstring."""
from collections import OrderedDict
from typing import Dict
from typing import Tuple
from typing import Union
import uuid

from immutabledict import immutabledict
from labware_domain_models import LabwareDefinition
from mantarray_file_manager import MANTARRAY_SERIAL_NUMBER_UUID
from mantarray_file_manager import METADATA_UUID_DESCRIPTIONS
from mantarray_file_manager import PLATE_BARCODE_UUID
from mantarray_file_manager import TISSUE_SAMPLING_PERIOD_UUID
from mantarray_file_manager import UTC_BEGINNING_RECORDING_UUID
from mantarray_file_manager import WELL_NAME_UUID
from mantarray_waveform_analysis import AMPLITUDE_UUID
from mantarray_waveform_analysis import AUC_UUID
from mantarray_waveform_analysis import BESSEL_LOWPASS_10_UUID
from mantarray_waveform_analysis import BUTTERWORTH_LOWPASS_30_UUID
from mantarray_waveform_analysis import CENTIMILLISECONDS_PER_SECOND
from mantarray_waveform_analysis import TWITCH_FREQUENCY_UUID
from mantarray_waveform_analysis import TWITCH_PERIOD_UUID
from mantarray_waveform_analysis import WIDTH_UUID

try:  # adapted from https://packaging.python.org/guides/single-sourcing-package-version/
    from importlib import metadata
except ImportError:  # pragma: no cover
    # Running on pre-3.8 Python; use importlib-metadata package
    import importlib_metadata as metadata  # type: ignore # Eli (9/1/20): for some reason mypy is giving weird errors for this
PACKAGE_VERSION = metadata.version("curibio.sdk")  # type: ignore # Eli (9/1/20): for some reason mypy is giving weird errors for this

TWENTY_FOUR_WELL_PLATE = LabwareDefinition(row_count=4, column_count=6)

MICROSECONDS_PER_CENTIMILLISECOND = 10
METADATA_EXCEL_SHEET_NAME = "metadata"
METADATA_RECORDING_ROW_START = 0
METADATA_INSTRUMENT_ROW_START = METADATA_RECORDING_ROW_START + 4
METADATA_OUTPUT_FILE_ROW_START = METADATA_INSTRUMENT_ROW_START + 6

CONTINUOUS_WAVEFORM_SHEET_NAME = "continuous-waveforms"
AGGREGATE_METRICS_SHEET_NAME = "aggregate-metrics"
PER_TWITCH_METRICS_SHEET_NAME = "per-twitch-metrics"
NUMBER_OF_PER_TWITCH_METRICS = 18
SNAPSHOT_CHART_SHEET_NAME = "continuous-waveform-snapshots"
FULL_CHART_SHEET_NAME = "full-continuous-waveform-plots"
TWITCH_FREQUENCIES_CHART_SHEET_NAME = "twitch-frequencies-plots"
FORCE_FREQUENCY_RELATIONSHIP_SHEET = "force-frequency-relationship"

INTERPOLATED_DATA_PERIOD_SECONDS = 1 / 100
INTERPOLATED_DATA_PERIOD_CMS = (
    INTERPOLATED_DATA_PERIOD_SECONDS * CENTIMILLISECONDS_PER_SECOND
)
TSP_TO_DEFAULT_FILTER_UUID = (
    {  # Tissue Sampling Period (centi-milliseconds) to default Pipeline Filter UUID
        960: BESSEL_LOWPASS_10_UUID,
        160: BUTTERWORTH_LOWPASS_30_UUID,
    }
)

DEFAULT_CELL_WIDTH = 64
CHART_HEIGHT = 300
CHART_BASE_WIDTH = 120
CHART_HEIGHT_CELLS = 15
CHART_FIXED_WIDTH_CELLS = 8
CHART_FIXED_WIDTH = DEFAULT_CELL_WIDTH * CHART_FIXED_WIDTH_CELLS
PEAK_VALLEY_COLUMN_START = 100
CHART_WINDOW_NUM_SECONDS = 10
CHART_WINDOW_NUM_DATA_POINTS = (
    CHART_WINDOW_NUM_SECONDS / INTERPOLATED_DATA_PERIOD_SECONDS
)
SECONDS_PER_CELL = 2.5

CALCULATED_METRIC_DISPLAY_NAMES: Dict[
    uuid.UUID, Union[str, Tuple[int, str]]
] = OrderedDict(
    [
        (TWITCH_PERIOD_UUID, "Twitch Period (seconds)"),
        (TWITCH_FREQUENCY_UUID, "Twitch Frequency (Hz)"),
        (AMPLITUDE_UUID, "Twitch Amplitude"),
        (WIDTH_UUID, (50, "Twitch Width 50 (FWHM) (seconds)")),
        (AUC_UUID, "Twitch Area Under the Curve (AUC)"),
    ]
)
ALL_FORMATS = immutabledict({"CoV": {"num_format": "0.00%"}})

TWITCHES_POINT_UP_UUID = uuid.UUID("97f69f56-f1c6-4c50-8590-7332570ed3c5")
INTERPOLATION_VALUE_UUID = uuid.UUID("466d0131-06b7-4f0f-ba1e-062a771cb280")
METADATA_UUID_DESCRIPTIONS.update(
    {
        TWITCHES_POINT_UP_UUID: "Flag indicating whether or not the twitches in the data point up or not",
        INTERPOLATION_VALUE_UUID: "Desired value for optical well data interpolation",
    }
)

EXCEL_OPTICAL_METADATA_CELLS = {
    WELL_NAME_UUID: "E2",
    UTC_BEGINNING_RECORDING_UUID: "E3",
    PLATE_BARCODE_UUID: "E4",
    TISSUE_SAMPLING_PERIOD_UUID: "E5",
    TWITCHES_POINT_UP_UUID: "E6",
    MANTARRAY_SERIAL_NUMBER_UUID: "E7",
    INTERPOLATION_VALUE_UUID: "E8",
}


src/curibio/sdk/files.py
# -*- coding: utf-8 -*-
"""Classes and functinos for finding and managing files."""
import inspect
import os
from typing import Dict
from typing import List

import h5py
from nptyping import NDArray
import numpy as np

PATH_OF_CURRENT_FILE = os.path.dirname((inspect.stack()[0][1]))


def get_unique_files_from_directory(directory: str) -> List[str]:
    """Obtain a list of all unique h5 files in the current directory.

    Args:
        directory: the master folder for which all h5 files reside

    Returns:
        A list of the file paths for all the h5 files in the directory.
    """
    unique_files: List[str] = []

    for path, _, files in os.walk(directory):
        for name in files:
            unique_files.append(os.path.join(path, name))

    return unique_files


def get_specified_files(
    search_criteria: str, criteria_value: str, unique_files: List[str]
) -> Dict[str, Dict[str, List[str]]]:
    """Obtain a subset of all the h5 files based on search criteria from user.

    Args:
        search_criteria: A str representing a UUID to a piece a metadata to filter results

    Returns:
        a dictionary of the Plate Recordings that fall under the specified search criteria.
    """
    # unique_files: List[str] = get_unique_files_from_directory(PATH_OF_CURRENT_FILE)

    value_dict: Dict[str, List[str]] = {}
    full_dict: Dict[str, Dict[str, List[str]]] = {}
    plate_recording_list: List[str] = []

    for file in unique_files:
        well = WellFile(file)
        if search_criteria == "Well Name" and well.get_well_name() == criteria_value:
            plate_recording_list.append(file)
        if (
            search_criteria == "Plate Barcode"
            and well.get_plate_barcode() == criteria_value
        ):
            plate_recording_list.append(file)
        if search_criteria == "User ID" and well.get_user_account() == criteria_value:
            plate_recording_list.append(file)
        if (
            search_criteria == "Account ID"
            and well.get_customer_account() == criteria_value
        ):
            plate_recording_list.append(file)
        if (
            search_criteria == "Mantarray Serial Number"
            and well.get_mantarray_serial_number() == criteria_value
        ):
            plate_recording_list.append(file)

    value_dict = {criteria_value: plate_recording_list}
    full_dict = {search_criteria: value_dict}

    return full_dict


class WellFile:
    """Wrapper around an H5 file for a single well of data.

    Args:
        file_name: The path of the H5 file to open.

    Attributes:
        _h5_file: The opened H5 file object.
    """

    def __init__(self, file_name: str) -> None:
        self._h5_file: h5py._hl.files.File = h5py.File(
            file_name, "r",
        )

    def get_well_name(self) -> str:
        return str(self._h5_file.attrs["Well Name"])

    def get_well_index(self) -> int:
        return int(self._h5_file.attrs["Well Index (zero-based)"])

    def get_plate_barcode(self) -> str:
        return str(self._h5_file.attrs["cf60afef-a9f0-4bc3-89e9-c665c6bb6941"])

    def get_user_account(self) -> str:
        return str(self._h5_file.attrs["7282cf00-2b6e-4202-9d9e-db0c73c3a71f"])

    def get_customer_account(self) -> str:
        return str(self._h5_file.attrs["4927c810-fbf4-406f-a848-eba5308576e6"])

    def get_mantarray_serial_number(self) -> str:
        return str(self._h5_file.attrs["83720d36-b941-4d85-9b39-1d817799edd6"])

    # def get_UTC_begin_recording(self) -> str:
    #     return str(self._h5_file.attrs["d2449271-0e84-4b45-a28b-8deab390b7c2"])

    def get_numpy_array(self) -> NDArray[2, float]:
        """Return the data (tissue sensor vs time)."""
        time_step = 8 * 1.2e-3  # tissue sample rate is just over 100Hz
        tissue_data = self._h5_file["tissue_sensor_readings"]

        times = np.arange(len(tissue_data)) * time_step
        len_time = len(times)

        data = np.zeros((2, len_time))
        for i in range(len_time):
            data[0, i] = times[i]
            data[1, i] = tissue_data[i]

        return data

    def get_voltage_array(self) -> NDArray[2, float]:
        """Return the voltage vs time data."""
        time_step = 8 * 1.2e-3  # tissue sample rate is just over 100Hz
        vref = 3.3  # ADC reference voltage
        least_significant_bit = vref / 2 ** 23  # ADC quantization step
        gain = 1  # ADC gain
        tissue_data = self._h5_file["tissue_sensor_readings"]

        times = np.arange(len(tissue_data)) * time_step
        len_time = len(times)

        voltages = []
        for this_tissue_data in tissue_data:
            voltages.append(1e3 * (least_significant_bit * this_tissue_data) / gain)

        voltage_data = np.zeros((2, len_time))
        for i in range(len_time):
            voltage_data[0, i] = times[i]
            voltage_data[1, i] = voltages[i]

        return voltage_data


class PlateRecording:
    """Wrapper around 24 WellFiles fpr a single plate of data.

    Args:
        file_paths: A list of all the file paths for each h5 file to open.

    Attributes:
        _files_ : WellFiles of all the file paths provided.
    """

    def __init__(self, file_paths: List[str]) -> None:
        self._files_: List[str] = file_paths

    def get_wellfile_names(self) -> List[str]:
        well_files = []
        for well in self._files_:
            well_files.append(WellFile(well).get_well_name())
        return well_files

    def get_combined_csv(self) -> None:
        data = np.zeros((len(self._files_) + 1, 12315))
        for i, well in enumerate(self._files_):
            well_data = WellFile(well).get_numpy_array()
            data[0, :] = well_data[0, :]
            data[i + 1, :] = well_data[1, :]

        my_local_path_data = os.path.join(PATH_OF_CURRENT_FILE, "PlateRecording.csv")
        np.savetxt(my_local_path_data, data, delimiter=",", fmt="%d")
